### DISCUSSION TITLE
SCCP support

### DISCUSSION THREAD

**Discussion by dstogov** [SeaOfNodes/Simple#144] [Ideas]
Do you plan to implement Sparse Conditional Constants Propagation?
It took me a while to understand that the iterative optimization pass doesn't do it at all.
The example from "Combining Analyses, Combining Optimizations"?   

```java
        Parser parser = new Parser(
                """
int x = 1;
int i = 1;
while(i < 10) {
    if (x != 1) {
    	x = 2;
    }
    i = i + 1;
}
return x;
                """);
        StopNode stop = parser.parse().iterate();
        assertEquals("return 1", stop.toString());
```

**Comment by cliffclick** [comment]
Yes, its in the plans.  Honestly should be really tiny amount of work at this point; the compiler is geared to support it.

**Comment by dstogov** [comment]
Did you see  "A sparse algorithm for predicated global value numbering" by Karthik Gargi ([paper](https://dl.acm.org/doi/10.1145/512529.512536) and [slides]( https://www.sable.mcgill.ca/~hendren/621/karthikhandouts.pdf)). The algorithm may perform combination of SCCP and GVN, and it's significantly simpler than your algorithm from "Combining Analyses, Combining Optimizations".  I plan to try it in my project.

**Reply by cliffclick** [reply]
Yes, good luck with your implementation.
I will point out that some of the examples at the bottom are flawed, in that normal dominating test - as has been done by many compilers including C2 for many decades - cleans out the examples, such that they never reach a later analysis.  C2 will clean such things out while parsing bytecodes and building SSA, i.e. before any other passes.  This makes reading improvement numbers a little gray: no optimization runs in a vacuum, and easy/fast/cheap earlier phases may "eat the heart" out of any gains made in later optimizations.
